src/main/java/com/google/sps/data/OrderHandler.java

```diff
@@ -0,0 +1,109 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.maps.errors.ApiException;
+import java.io.IOException;
+import java.lang.InterruptedException;
+import java.lang.*;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+/** 
+ * Class used for creating and storing orders.
+ */
+public class OrderHandler {
+  private PathFinder pathFinder;
+
+  public OrderHandler(PathFinder pathFinder) {
+    this.pathFinder = pathFinder;
+  }
+
+  /** 
+   * Creates an "Order" Entity and adds it to datastore. Each order contains a set of books and the
+   * library that has all the requested books.
+   */
+  private void addOrderToDatastore(LibraryPoint library, ArrayList<String> bookIds, String userId, Point address) {
+    Entity order = new Entity("Order");
+    order.setProperty("libraryId", (int)library.getLibraryId());
+    order.setProperty("libraryLatitude", library.latitude);
+    order.setProperty("libraryLongitude", library.longitude);
+    order.setProperty("recipientLatitude", address.latitude);
+    order.setProperty("recipientLongitude", address.longitude);
+    order.setProperty("status", "ADDED");
+    order.setProperty("area", library.getArea());
+    order.setProperty("userId", userId);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(order);
+  }
+
+  /**
+   * Given a user's Id, the shipping address and an array of book ids, it creates a set of orders
+   * and adds them to datastore.
+   */
+  public boolean makeOrders(String userId, Point address, List<String> bookIds) throws ApiException, IOException, InterruptedException {
+    HashMap <LibraryPoint, ArrayList<String>> libraryBookIds = new HashMap<>();
+
+    for (String bookId : bookIds) {
+      Entity closestLibrary = getClosestLibrary(
+          BooksManager.getLibrariesForBook(bookId, /** setLimit = */false), address);
+      if (closestLibrary == null) {
+        // there is no library that has bookId, thus the order can't be made
+        return false;
```

Comment:
So if I had an order for say 10 books, and just one cannot be made, the whole thing is rejected.

That is fine, but how does the calling coded know which book was rejected?

Could it make more sense to return a collection back to the user so that we have a true/false status for each individual book, with an overall status for the entire order?  You might need to make an `OrderStatus` class to return that info.

Reply:
You're right, I changed the function to return a Collection with the book ids of the rejected books. I can change it to return a Collection of `OrderStatus` objects if that is better, but the status of books can be inferred knowing which books are rejected.

Reply:
A list of rejected books sounds good to me.